taxonomy/tests/test_api.py

```diff
@@ -0,0 +1,346 @@
+"""
+Tests for taxonomy app
+"""
+
+# pylint: disable=no-member
+# pylint: disable=too-many-instance-attributes
+# pylint: disable=invalid-name
+
+from django.test.testcases import TestCase
+from django.contrib.auth.models import User
+from django.db.utils import IntegrityError
+
+from taxonomy.models import (
+    Vocabulary,
+    Term,
+)
+from learningresources.models import (
+    LearningResource,
+    LearningResourceType,
+    Repository,
+    Course,
+)
+from taxonomy.api import (
+    add_term_to_learning_resource,
+    add_type_for_vocabulary,
+    create_term,
+    create_vocabulary,
+    delete_term,
+    delete_vocabulary,
+    get_learning_resources_for_term,
+    get_term,
+    get_terms_for_learning_resource,
+    get_types_for_vocabulary,
+    get_vocabularies_for_type,
+    get_vocabulary,
+    remove_term_from_learning_resource,
+    remove_type_from_vocabulary,
+    TaxonomyAPIException,
+)
+
+
+class TestApi(TestCase):
+    """Tests for taxomony API"""
+
+    def setUp(self):
+        super(TestApi, self).setUp()
+        self.user = User.objects.create(username="test")
+
+        self.repository = Repository.objects.create(
+            create_date="2014-08-08",
+            created_by=self.user,
+        )
+
+        self.vocabulary = Vocabulary.objects.create(
+            repository=self.repository,
+            required=False,
+            weight=100,
+        )
+
+        self.learning_resource_type1 = LearningResourceType.objects.create(
+            name="Type1",
+        )
+        self.learning_resource_type2 = LearningResourceType.objects.create(
+            name="Type2",
+        )
+
+        self.course = Course.objects.create(
+            repository=self.repository,
+            imported_by=self.user,
+        )
+
+        self.learning_resource = LearningResource.objects.create(
+            course=self.course,
+            learning_resource_type=self.learning_resource_type1,
+        )
+
+        self.term = Term.objects.create(
+            vocabulary=self.vocabulary,
+            weight=4,
+        )
+
+    def test_add_term_to_learning_resource(self):
+        """
+        Test add_term_to_learning_resource
+        """
+        self.assertEquals(0, self.learning_resource.terms.count())
+        add_term_to_learning_resource(self.learning_resource.id, self.term.id)
+        self.assertEquals([self.term],
+                          list(self.learning_resource.terms.all()))
+        add_term_to_learning_resource(self.learning_resource.id, self.term.id)
+
+        # duplicate term is ignored
+        self.assertEquals([self.term],
+                          list(self.learning_resource.terms.all()))
+
+        self.assertRaises(LearningResource.DoesNotExist,
```

Comment:
These are good tests, but if these cases happened in production it would probably be a runtime error, resulting in a 500 response to the user.

The suggestions I made about using `full_clean` within the API calls will help with this.